File: leetcode2/2medium/강병호/Q2244.py
class Solution:
    def minimumRounds(self, tasks: list[int]) -> int:
        key_tasks = {}

        for task in tasks:
            if not task in key_tasks:
                key_tasks[task] = 1
                continue
            key_tasks[task] += 1
        
        # 브루트 포스
        key_list = key_tasks.keys()
        # print(f"key_list : {key_list}")
        result = 0

        for key in key_list:
            # print(f"key : {key_tasks[key]}")
            while key_tasks[key] > 1:
                if key_tasks[key] == 4:
                    key_tasks[key] -= 4
                    result += 2
                elif key_tasks[key] >= 3:
                    key_tasks[key] -= 3
                    result += 1
                else:
                    key_tasks[key] -= 2
                    result += 1
            
            if key_tasks[key] == 1:
                result = -1
                break
        
        return result
    
    # def recur(self, value:int) -> bool:

if __name__ == "__main__":
    sol = Solution()
    print(sol.minimumRounds([5,5,5,5]))

    print(sol.minimumRounds([2, 3, 3]))
    print(sol.minimumRounds([2,2,3,3,2,4,4,4,4,4]))

File: leetcode2/2medium/김대원/Q2244.java
// 1. 아이디어 : 3으로 나누어지는지 판단하고, 경우의 수로 라운드 카운팅
// 2. 시간복잡도 : O(N)
// 3. 자료구조/알고리즘 : 그리디

class Solution {
    public int minimumRounds(int[] tasks) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int x : tasks) map.put(x, map.getOrDefault(x, 0) + 1);

        int round = 0;
        for (int k : map.keySet()) {
            int task = map.get(k);

            if (task == 1) return -1;

            if (task % 3 == 0) round += task / 3;
            else if (task % 3 == 1) round += (task / 3) - 1 + 2;
            else if (task % 3 == 2) round += task / 3 + 1;
        }

        return round;
    }
}